### DISCUSSION TITLE
onError method  not getting invoked on LoggingProducerListener exception

### DISCUSSION THREAD

**Discussion by shravani154** [spring-projects/spring-kafka#2303] [Q&A]
I have a logic in onError method which needs to be invoked whenever there is a LoggingProducerListener exception. I are not able to catch this exception through try catch block also. I need to know why this exception occurs and how to handle/catch these exceptions. Below is the exception I am encountering. 

2022-06-10 15:22:01.126 ERROR 1 --- [ad | producer-1] o.s.k.support.LoggingProducerListener    : Exception thrown when sending a message with key='null' and payload='{123, 34, 116, 114, 97, 99, 101, 115, 34, 58, 32, 123, 34, 115, 101, 114, 118, 105, 99, 101, 78, 97,...' to topic

**Comment by artembilan** [comment]
What you are talking about is not a thrown exception, it is a log message under ERROR category.
See the logic of that `LoggingProducerListener`:
```
	public void onError(ProducerRecord<K, V> record, @Nullable RecordMetadata recordMetadata, Exception exception) {
		this.logger.error(exception, () -> {
			StringBuffer logOutput = new StringBuffer();
			logOutput.append("Exception thrown when sending a message");
			if (this.includeContents) {
				logOutput.append(" with key='")
					.append(keyOrValue(record.key()))
					.append("'")
					.append(" and payload='")
					.append(keyOrValue(record.value()))
					.append("'");
			}
			logOutput.append(" to topic ").append(record.topic());
			if (record.partition() != null) {
				logOutput.append(" and partition ").append(recordMetadata != null
						? recordMetadata.partition()
						: record.partition());
			}
			logOutput.append(":");
			return logOutput.toString();
		});
	}
```

So, probably better to inject your own `ProducerListener` with that respective `onError()` impl.
But again: that `try..catch` is done in the Kafka Client where they call a `Callback` provided, which in our cases just delegates to the provided  `ProducerListener`.

In two words: the `LoggingProducerListener` does not throw exceptions to catch: it just logs them as ERRORs.

**Reply by shravani154** [reply]
Why do we get this exception frequently and how to avoid it?

**Reply by artembilan** [reply]
What exception, please?
Isn't this a new question though?
Looks like you have switched context and is already not about `LoggingProducerListener`...

**Reply by shravani154** [reply]
Sorry for the confusion.
I am reffering to LoggingProducerListener error. I am trying to understand why do we get this error and are there any ways to avoid this?

2022-06-10 15:22:01.126 ERROR 1 --- [ad | producer-1] o.s.k.support.LoggingProducerListener : Exception thrown when sending a message with key='null' and payload='{123, 34, 116, 114, 97, 99, 101, 115, 34, 58, 32, 123, 34, 115, 101, 114, 118, 105, 99, 101, 78, 97,...' to topic

**Reply by garyrussell** [reply]
You need to look at the stack trace in the to determine the cause; show the complete log.

**Reply by artembilan** [reply]
You show just a beginning of the message from that `LoggingProducerListener`:
```
		this.logger.error(exception, () -> {
			StringBuffer logOutput = new StringBuffer();
			logOutput.append("Exception thrown when sending a message");
```

But you definitely don't show what is that `exception`. Therefore no way to help you so far.

**Reply by shravani154** [reply]
Here is the full message:
2022-06-13 14:59:47.458 ERROR 1 --- [ad | producer-1] o.s.k.support.LoggingProducerListener    : Exception thrown when sending a message with key='null' and payload='{123, 34, 116, 114, 97, 99, 101, 115, 34, 58, 32, 123, 34, 115, 101, 114, 118, 105, 99, 101, 78, 97,...' to topic ***:

org.apache.kafka.common.errors.PolicyViolationException: Request parameters do not satisfy the configured policy.

**Reply by artembilan** [reply]
> Request parameters do not satisfy the configured policy.

Sounds like some Apache Kafka security config violation.

Consider to ask such a question in broader Kafka community.
This one is out of Spring for Apache Kafka scope.

**Reply by garyrussell** [reply]
Perhaps there are more clues in the server logs. If you can't figure it out from that, ask about this on Stack Overflow or on one the Kafka mailing lists: https://kafka.apache.org/contact

**Reply by shravani154** [reply]
Another question regarding LoggingProducerListener.
 I am using cloud stream binders in my application and have multiple binders. I have my own Producer Listener in my application which logs on success and on error. In this scenario, I am not able to see onsuccess and on error messages if I use multiple binders. If I have a single binder the class is getting called, The issue is only when I have multiple binders. Is there any configuration I should set up? 

Thanks In advance.

**Reply by garyrussell** [reply]
See the documentation - https://docs.spring.io/spring-cloud-stream/docs/current/reference/html/spring-cloud-stream.html#binder-customizer

You need to use a binder customizer instead when there are multiple binders.

**Reply by KhaiNguyenDuc** [reply]
Thanks, I have the same problem
